src/openms/include/OpenMS/KERNEL/MSExperiment.h

```diff
@@ -392,6 +392,395 @@ namespace OpenMS
+  return aggregate(mz_rt_ranges, ms_level, SumIntensityReduction());
+}
+
+/**
+ * @brief Extracts extracted ion chromatograms (XICs) from the MSExperiment.
+ *
+ * This function takes a vector of mz_rt_ranges, an ms_level, and a MzReductionFunctionType
+ * and extracts the XICs from the MSExperiment based on the given parameters.
+ *
+ * @param mz_rt_ranges A vector of pairs of RangeMZ and RangeRT representing the m/z and retention time ranges.
+ * @param ms_level The MS level of the spectra to consider.
+ * @param func_mz_reduction The MzReductionFunctionType used to reduce the m/z values.
+ *
+ * @return A vector of MSChromatogram objects representing the extracted XICs.
+ */
+template<class MzReductionFunctionType>
+std::vector<MSChromatogram> extractXICs(
+    std::vector<std::pair<RangeMZ, RangeRT>>& mz_rt_ranges,
+    unsigned int ms_level,
+    MzReductionFunctionType func_mz_reduction) const
+{
+    // Early exit if there are no ranges
+    if (mz_rt_ranges.empty()) 
+    {
+      // likely an error, but we return an empty vector instead of throwing an exception for now
+      return {};
+    }
+
+    // Sort mz_rt_ranges by ascending MinMZ and descending MaxMZ
+    std::sort(mz_rt_ranges.begin(), mz_rt_ranges.end(), 
+        [](const auto& a, const auto& b) 
+        {
+          if (a.first.getMinMZ() != b.first.getMinMZ()) 
+          {
+            return a.first.getMinMZ() < b.first.getMinMZ();
+          }
+          return -a.first.getMaxMZ() < -b.first.getMaxMZ();
+        });
+
+    // Create a view of the spectra with given MS level
+    std::vector<std::reference_wrapper<const MSSpectrum>> spectra_view;
+    spectra_view.reserve(spectra_.size());
+    std::copy_if(spectra_.begin(), spectra_.end(), 
+                 std::back_inserter(spectra_view),
+                 [ms_level](const auto& spec) { 
+                     return spec.getMSLevel() == ms_level; 
+                 });
+
+    // Early exit if there are no spectra with the given MS level
+    if (spectra_view.empty()) { // could be valid use or an error -> we return an empty vector
+      return {};
+    }
+
+    // Get the indices of the spectra covered by the RT ranges by considering the MS level
+    // If start and stop are the same, the range is empty
+    auto getCoveredSpectra = [](
+        const std::vector<std::reference_wrapper<const MSSpectrum>>& spectra_view, 
+        std::vector<std::pair<RangeMZ, RangeRT>>& mz_rt_ranges) 
+        ->  std::vector<std::pair<size_t, size_t>>
+      {
+        std::vector<std::pair<size_t, size_t>> res;
+        res.reserve(mz_rt_ranges.size());
+        for (const auto & mz_rt : mz_rt_ranges) 
+        {
+          auto start_it = std::lower_bound(spectra_view.begin(), spectra_view.end(), mz_rt.second.getMin(), 
+            [](const auto& spec, double rt) 
+            { return spec.get().getRT() < rt; });
+
+          auto stop_it = std::upper_bound(spectra_view.begin(), spectra_view.end(), mz_rt.second.getMax(), 
```

Comment:
For smaller ranges, a linear search to the end is faster. Consider this.

Reply:
> For smaller ranges, a linear search to the end is faster. Consider this.

True, but it's not guaranteed. And in the worst case you have runtime O(N). Binary search always does O(logN), so use that :)

> If you keep the binary search, why not start searching from the start it? Might save you one comparison or so on average.

Usually not a good idea since the elements traversed during BS are already in cache when using [begin, end). If you use another `begin`, then the cache is cold. This should be true for small and large resulting ranges, thus its always better to stay with the full range.

Reply:
Ok cool. Convinced

Reply:
I think this should be optimized for sparse mz ranges as this will be our main use case for now. Let's get something in and optimize later :)